cpp/src/arrow/array/builder_nested.h

```diff
@@ -329,9 +329,7 @@ class ARROW_EXPORT MapBuilder : public ArrayBuilder {
   /// the Map to be built as a list of struct values.
   ArrayBuilder* value_builder() const { return list_builder_->value_builder(); }
 
-  std::shared_ptr<DataType> type() const override {
-    return map(key_builder_->type(), item_builder_->type(), keys_sorted_);
-  }
+  std::shared_ptr<DataType> type() const override { return type_; }
```

Comment:
Note that the `map` construction I deleted is already used to set the type when it isn't explicitly passed.

https://github.com/apache/arrow/blob/e734856676a00335f3dfe79899a995d87286f5a9/cpp/src/arrow/array/builder_nested.cc#L52-L55

Reply:
Hmm... unfortunately, I don't think this is right. Some child builders can actually change type over time (for example the index width of a dictionary may increase depending on the dictionary size). So we have to keep the dynamic computation in the `type()` method.

Reply:
Missed this. That makes sense. I think the problem is that those types don't have field names, so maybe the solution is to dynamically add back the field names within this method.

Reply:
@pitrou I've updated this part to rebuild the type with the child builders and the saved field names.